server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1343,13 +1346,9 @@ private void innerOpenEngineAndTranslog() throws IOException {
         final String translogUUID = store.readLastCommittedSegmentsInfo().getUserData().get(Translog.TRANSLOG_UUID_KEY);
         final long globalCheckpoint = Translog.readGlobalCheckpoint(translogConfig.getTranslogPath(), translogUUID);
         replicationTracker.updateGlobalCheckpointOnReplica(globalCheckpoint, "read from translog checkpoint");
-
-        assertMaxUnsafeAutoIdInCommit();
-
-        final long minRetainedTranslogGen = Translog.readMinTranslogGeneration(translogConfig.getTranslogPath(), translogUUID);
-        store.trimUnsafeCommits(globalCheckpoint, minRetainedTranslogGen, config.getIndexSettings().getIndexVersionCreated());
-
-        createNewEngine(config);
+        synchronized (mutex) {
+            engineHolder.activateNewEngine(createNewEngine(config));
```

Comment:
We need to create a new engine under mutex, otherwise, we might clean up an index commit which is being snapshotted by `IndexShard#snapshotStoreMetadata` (without an engine).